### PR TITLE
fixes to posptocessing, kilosort, and multicomp

### PR DESCRIPTION
**Pull request by alejoe91**
- updated export to phy function with improved pca extraction and group settings (working with latest `gloo` phy branch)

-  fixed channel mismatch in get_unit_max_channel

- for kilosort and kilosort2: if location is not given, linear location is assumed

- in multisorting comparison: `compare_spiketrains` checks that spikes have not be labeled as TP before

### CODE REVIEW

File: spiketoolkit/comparison/comparisontools.py [hunk 1]
```diff
@@ -475,8 +475,9 @@ def compare_spike_trains(spiketrain1, spiketrain2, delta_frames=10):
     for sp_i, n_sp in enumerate(spiketrain1):
         matches = (np.abs(spiketrain2.astype(int) - n_sp) <= delta_frames // 2)
         if np.sum(matches) > 0:
-            lab_st1[sp_i] = 'TP'
-            lab_st2[np.where(matches)[0][0]] = 'TP'
+            if lab_st1[sp_i] != 'TP' and lab_st2[np.where(matches)[0][0]] != 'TP':
+                lab_st1[sp_i] = 'TP'
+                lab_st2[np.where(matches)[0][0]] = 'TP'
```
Comment (samuelgarcia) [inline spiketoolkit/comparison/comparisontools.py]: Why this logic is differents from line 265 in do_score_labels ?

File: spiketoolkit/postprocessing/postprocessing_tools.py [hunk 2]
```diff
@@ -365,8 +366,9 @@ def get_unit_max_channel(recording, sorting, unit_ids=None, grouping_property=No
                                        ms_before=ms_before, ms_after=ms_after,  grouping_property=grouping_property,
                                        compute_property_from_recording=compute_property_from_recording,
                                        verbose=verbose)
-        max_channel = np.unravel_index(np.argmax(np.abs(template)),
-                                       template.shape)[0]
+        max_chan = np.unravel_index(np.argmax(np.abs(template)),
+                                    template.shape)[0]
+        max_channel = recording.get_channel_ids()[max_chan]
```
Comment (samuelgarcia) [inline spiketoolkit/postprocessing/postprocessing_tools.py]: Here I would explicitly use **max_chan_index** and **max_chan_id** in the code in case of the variable will be used further.

File: spiketoolkit/postprocessing/postprocessing_tools.py [hunk 3]
```diff
@@ -480,20 +482,14 @@ def compute_pca_scores(recording, sorting, unit_ids=None, n_comp=3, by_electrode
     if verbose:
         print("Fitting PCA of %d dimensions on %d waveforms" % (n_comp, len(all_waveforms)))
 
-    pca = PCA(n_components=n_comp, whiten=True)
-    pca.fit_transform(all_waveforms)
-    scores = pca.transform(all_waveforms)
+    pca = PCA(n_components=n_comp, whiten=False)
+    pca.fit(all_waveforms)
```
Comment (samuelgarcia) [inline spiketoolkit/postprocessing/postprocessing_tools.py]: maybe the **whitten** could be expose at the method level.
In case of already whitten signals.

File: spiketoolkit/preprocessing/remove_bad_channels.py [hunk 4]
```diff
@@ -57,14 +56,15 @@ def _initialize_subrecording_extractor(self):
             traces = self._recording.get_traces(start_frame=start_frame, end_frame=end_frame)
             stds = np.std(traces, axis=1)
             bad_channels = [ch for ch, std in enumerate(stds) if std > self._bad_threshold * np.median(stds)]
+            print('Automatically removing channels:', bad_channels)
```
Comment (samuelgarcia) [inline spiketoolkit/preprocessing/remove_bad_channels.py]: With a self.verbose ? No ?

File: spiketoolkit/sorters/kilosort/kilosort.py [hunk 5]
```diff
@@ -155,26 +155,27 @@ def _setup_recording(self, recording, output_folder):
...
+            for i_ch, ch in enumerate(recording.get_channel_ids()):
+                recording.set_channel_property(ch, 'location', [0, i_ch])
+        positions = np.array([recording.get_channel_property(chan, 'location') for chan in recording.get_channel_ids()])
+        if electrode_dimensions is None:
+            kilosort_channelmap = ''.join(kilosort_channelmap
+                                          ).format(nchan,
+                                                   list(positions[:, 0]),
+                                                   list(positions[:, 1]),
+                                                   groups,
+                                                   sample_rate)
```
Comment (samuelgarcia) [inline spiketoolkit/sorters/kilosort/kilosort.py]: This will make the generated file with MEArec not working because you use dim 1 and 2.
No ?
We shoucld write somewhere that if 3D in recording but sorter need 2D then the first 2 dimes are taken magically.

**Comment by samuelgarcia** [comment]
The trick you made for KS, about the 2 dim, should be share across all sorter in BaseSorter.
Adding a attr in BaseSroter than handle this and make the trick automatically should be done.

File: spiketoolkit/postprocessing/postprocessing_tools.py (hunk 2 above)
Comment (alejoe91) [reply spiketoolkit/postprocessing/postprocessing_tools.py]: ok

File: spiketoolkit/comparison/comparisontools.py (hunk 1 above)
Comment (alejoe91) [reply spiketoolkit/comparison/comparisontools.py]: we should chnge it there as well. This ensures that the number of TP in both spike trains is the same.

File: spiketoolkit/postprocessing/postprocessing_tools.py (hunk 3 above)
Comment (alejoe91) [reply spiketoolkit/postprocessing/postprocessing_tools.py]: sounds good

File: spiketoolkit/sorters/kilosort/kilosort.py (hunk 5 above)
Comment (alejoe91) [reply spiketoolkit/sorters/kilosort/kilosort.py]: you can use the electrode_dimension. if `electrode_dimension=[1,2]` then those dimensions are used

File: spiketoolkit/preprocessing/remove_bad_channels.py (hunk 4 above)
Comment (alejoe91) [reply spiketoolkit/preprocessing/remove_bad_channels.py]: ok :)

File: spiketoolkit/postprocessing/postprocessing_tools.py (hunk 3 above)
Comment (alejoe91) [reply spiketoolkit/postprocessing/postprocessing_tools.py]: BTW, I found out that the pca.scores returned by PCA are not the same as the projections of the waveforms on the different eigenvectors...I spent almost the entire day yesterday trying to understand why..In the end I just substituted the pca_scores calculation with the manual projection

File: spiketoolkit/sorters/kilosort/kilosort.py (hunk 5 above)
Comment (alejoe91) [reply spiketoolkit/sorters/kilosort/kilosort.py]: we can deal with it in another PR :)

Comment (samuelgarcia) [reply spiketoolkit/sorters/kilosort/kilosort.py]: Of course but in this futur PR, we will nede to detsroy this trick inside KS.
Lets make an issue to not forget it.

**Comment by alejoe91** [comment]
@samuelgarcia can I merge this?
Idk why it says some checks haven't completed..it looks passed on travis

**Comment by samuelgarcia** [comment]
Yes of course.
